File: Code/branches/OldGeometry/cpp/LinearAlgebra/ArnoldiMatrix.py
__id__       = "$Id: ArnoldiMatrix.py 339 2008-07-29 19:17:18Z jlconlin $"
__author__   = "$Author: jlconlin $"
__version__ = " $Revision: 339 $"
__date__     = "$Date: 2008-07-29 13:17:18 -0600 (Tue, 29 Jul 2008) $"

"""
This is an implentation of Arnoldi's method where the linear operator is just a
matrix.
"""

import sys

import numpy

import Arnoldi

class ArnoldiMatrix(Arnoldi.Arnoldi):

    def __init__(self, A):
        self._A = A
        Arnoldi.Arnoldi.__init__(self)

    def Av(self):
        """
        """
        v = self._V[:,self._k]
        self._v = numpy.dot(self._A, v)

    def Restart(self):
        """
        """
        self._Restart()
        print "Restart #: %4i" %(self._r)

    def Iteration(self):
        """
        """
        self._Iteration()
        print "\tIteration#: %4i" %(self._k)

def Diagonal(n):
    A = numpy.diag(numpy.arange(1,n+1),0)
    return A

def Standard(n):
    A = Diagonal(n)
    A[2,1] = 1
    return A

if __name__ == "__main__":
    n = 5
    kI = 2
    jI = 2

    A = Diagonal(n)
    v = numpy.ones(n)
    print "A:\n%s\nv:%s" %(A, v)

    ArnMat = ArnoldiMatrix(A)
    ArnMat.arnoldi(v,4)
#   ArnMat.iram(v,2,4,2)
    print "V:\n%s\nH:\n%s" %(ArnMat._V, ArnMat._H)

File: Code/branches/OldGeometry/cpp/LinearAlgebra/ArnoldiMatrix.cpp
/*ArnoldiMatrix.cpp
 $Author: jlconlin $
 $Id: ArnoldiMatrix.cpp 342 2008-08-04 19:55:07Z jlconlin $
 $Revision: 342 $
 $Date: 2008-08-04 13:55:07 -0600 (Mon, 04 Aug 2008) $

 This is the implementation of Arnoldi's method for a smiple Matrix.
 */

#include <iostream>
#include <iomanip>
#include <ctime>
#include <cmath>
#include "Arnoldi.h"
#include "ArnoldiMatrix.h"
#include "boost/format.hpp"
#include "boost/numeric/ublas/io.hpp"

#include "libs/numeric/bindings/lapack/test/utils.h"

using std::cout;
using std::endl;
using std::abs;
using boost::format;

Arnoldi::cVec_t ArnoldiMatrix::_Av(){
    return prod( _A, cColumn_t(_V, _k) );
}

/*_IRestart is a restart for IRAM.*/
void ArnoldiMatrix::_IRestart(cVec_t& v){
    ptime now( boost::date_time::microsec_clock<ptime>::local_time() );

    Arnoldi::_IRestart(v);
    if( _Verbose ){
        double duration = (now-_start).total_microseconds()/1000000.0;
        cout << format( "Restart #: %i, Eigvalue = %10.8f,") 
            %_R %_Eigenvalues( _Eigenvalues.size()-1 )
             << format( " res = %9.3G, time: %6.2G sec." ) %_residual %duration
              << endl;
    }
}

/*_ERestart is a restart for ERAM.*/
void ArnoldiMatrix::_ERestart(cVec_t& v){
    ptime now( boost::date_time::microsec_clock<ptime>::local_time() );

    Arnoldi::_ERestart(v);
    if( _Verbose ){
        double duration = (now-_start).total_microseconds()/1000000.0;
        cout << format( "Restart #: %i, Eigvalue = %10.8f,") 
            %_R %_Eigenvalues( _Eigenvalues.size()-1 )
             << format( " res = %9.3G, time: %6.2G sec." ) %_residual %duration
              << endl;
    }
}

void ArnoldiMatrix::_Iteration(cVec_t& v){
    ptime now( boost::date_time::microsec_clock<ptime>::local_time() );

    Arnoldi::_Iteration(v);
    if( _Verbose ){
        double duration = (now-_startR).total_microseconds()/1000000.0;
        cout << format( "\tIteration #: %i, time: %6.2G sec." )
                %_k %duration << endl;
    }
    if( abs( _H(_k,_k-1) ) < _Tolerance ){
        _stoppedI = true;
        if( _Verbose ){
            cout << format("Stopping because H[%i,%i] = %8.2E < %8.2E.") 
                %_k %(_k-1) %_H(_k,_k-1) %_Tolerance << endl;
        }
    }
}


